### ISSUE TITLE
npm not istalled in IDE

### ISSUE THREAD

**Issue by maybeec** [devonfw/devon-ide#179]
Category: Bug or enhancement
Bug

Severity: Low , Medium, High, Critical
High

Expected behaviour (In case of bug)
NPM should be installed within the IDE and be only used from the IDE's software directory (respectively IDE speicific node installation)

Actual/current behaviour (In case of bug)
npm is taken  from /c/program files/nodejs/npm if it is already installed globally on the system.
ng is also installed in the global installatino of nodejs on the system.

```
*** Setting up node ***
Starting installation of node.js to /c/Users/mbrunnli/Downloads/devon-ide-scripts-3.0.0-beta11.tar/software/node from https://nodejs.org/dist/v10.16.0/node-v10.16.0-win-x64.zip
Downloading node.js-v10.16.0-windows.zip from https://nodejs.org/dist/v10.16.0/node-v10.16.0-win-x64.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
 63 16.6M   63 10.5M    0     0  2909k      0  0:00:05  0:00:03  0:00:02 2909kInstalling extensions...
Installing extension 'devonfw.devonfw-extension-pack' v4.1.1...
100 16.6M  100 16.6M    0     0  2678k      0  0:00:06  0:00:06 --:--:-- 2543k
Running command: unzip -q '/c/Users/mbrunnli/Downloads/node.js-v10.16.0-windows.zip' -d '/c/Users/mbrunnli/Downloads/devon-ide-scripts-3.0.0-beta11.tar/updates/extracted/node'
Succeeded to run command unzip
Successfully extracted archive node.js-v10.16.0-windows.zip to updates/extracted
Running command: mv '/c/Users/mbrunnli/Downloads/devon-ide-scripts-3.0.0-beta11.tar/updates/extracted/node/node-v10.16.0-win-x64' '/c/Users/mbrunnli/Downloads/devon-ide-scripts-3.0.0-beta11.tar/software/node'
Succeeded to run command mv
Successfully installed node.js
To be fully functional please rerun 'devon' command to update your PATH properly.
Running command: '/c/Users/mbrunnli/Downloads/devon-ide-scripts-3.0.0-beta11.tar/software/node/node.exe' -v
v10.16.0
Succeeded to verify installation of node.js ('/c/Users/mbrunnli/Downloads/devon-ide-scripts-3.0.0-beta11.tar/software/node/node.exe')

*** Setting up npm ***
npm is already installed at /c/Program Files/nodejs/npm

*** Setting up ng ***
Running command: npm install -g @angular/cli@latest --unsafe
Extension 'devonfw.devonfw-extension-pack' v4.1.1 was successfully installed..2.6 checking installable status
C:\Program Files\nodejs\ng -> C:\Program Files\nodejs\node_modules\@angular\cli\bin\ng

> @angular/cli@8.2.1 postinstall C:\Program Files\nodejs\node_modules\@angular\cli
> node ./bin/postinstall/script.js

+ @angular/cli@8.2.1
added 240 packages from 185 contributors in 45.967s
Succeeded to install angular-cli (npm)
```

Steps to reproduce (In case of bug)
having nodejs installed on your system globally, start a new setup and have a look at the logs.

Description of issue / expected enhacement /Comments

Your environment - windows/Linux, Devonfw version , component

**Comment by markusschuh** [comment]
Issue title is a bit misleading. npm is installed as port of node.js, so the install location of npm is not the issue. But anyway, you are right for the location, where components are located and executed. This needs more consistency inside the scripts. They should never rely on any PATH set from externally but instead define the PATH intern - or execute any single call of an external tool with help of related function.

With respect to the `ng` setup, I'd suggest to relocate this to its own issue.

It looks to me, as if the main issue for the _right_ target location of such modules is more, that the folder structure concept of devonfw with respect to IDE may need some more strengthening / update. Where exactly should IDE instance specific modules be installed? One option could be ${DEVON_IDE_HOME}/sofware/node (which would need to reinstall any such modules on any later node.js update) Another option is  ${DEVON_IDE_HOME}/npm. In oder to overwrite a potential per user node.js config setup, this may need some additions / changes for the environment variables in devon.properties and the .npmrc template referenced there.

**Comment by markusschuh** [comment]
For completeness I have started the `devon ide setup` on my windows from a clean as possible environment. (No preinstalled tools in PATH and no tool related env vars)

Since the PATH is not updated after `node` setup, no `npm` can be found. Hence the `ng setup` fails.

```
*** Setting up node ***
Starting installation of node.js to /e/opt/devon-ide-3.0.0-beta11/software/node from https://nodejs.org/dist/v10.16.0/node-v10.16.0-win-x64.zip
Artifact already exists at /c/Users/marku/Downloads/node.js-v10.16.0-windows.zip
To force update please delete the file and run again.
Running command: unzip -q '/c/Users/marku/Downloads/node.js-v10.16.0-windows.zip' -d '/e/opt/devon-ide-3.0.0-beta11/updates/extracted/node'
Succeeded to run command unzip
Successfully extracted archive node.js-v10.16.0-windows.zip to updates/extracted
Running command: mv '/e/opt/devon-ide-3.0.0-beta11/updates/extracted/node/node-v10.16.0-win-x64' '/e/opt/devon-ide-3.0.0-beta11/software/node'
Succeeded to run command mv
Successfully installed node.js
To be fully functional please rerun 'devon' command to update your PATH properly.
Running command: '/e/opt/devon-ide-3.0.0-beta11/software/node/node.exe' -v
v10.16.0
Succeeded to verify installation of node.js ('/e/opt/devon-ide-3.0.0-beta11/software/node/node.exe')

*** Setting up npm ***
Version v10.16.0 of node.js is already installed.
Running command: '/e/opt/devon-ide-3.0.0-beta11/software/node/node.exe' -v
v10.16.0
Succeeded to verify installation of node.js ('/e/opt/devon-ide-3.0.0-beta11/software/node/node.exe')

*** Setting up ng ***
Running command: npm install -g @angular/cli@latest --unsafe
/e/opt/devon-ide-3.0.0-beta11/scripts/command/../functions: line 60: npm: command not found

******** ATTENTION ********
ERROR: Failed to install angular-cli (npm)
We are sorry for the inconvenience. Please check the above errors, resolve them and try again.
Exit code was 127
ERROR: command 'ng setup' failed with exit code 127
devon-ide environment variables have been set for E:\opt\devon-ide-3.0.0-beta11 (workspace main
```

After `devon ide setup` I have run `devon` again, as script output suggest several times during setup.
Now the `ng setup` succeeds and is directed into the ${DEVON_IDE_HOME}\location\node:

```
E:\opt\devon-ide-3.0.0-beta11>scripts\devon ng setup
Running command: npm install -g @angular/cli@latest --unsafe
E:\opt\devon-ide-3.0.0-beta11\software\node\ng -> E:\opt\devon-ide-3.0.0-beta11\software\node\node_modules\@angular\cli\bin\ng

> @angular/cli@8.2.2 postinstall E:\opt\devon-ide-3.0.0-beta11\software\node\node_modules\@angular\cli
> node ./bin/postinstall/script.js

+ @angular/cli@8.2.2
added 240 packages from 185 contributors in 23.227s
Succeeded to install angular-cli (npm)
devon-ide environment variables have been set for E:\opt\devon-ide-3.0.0-beta11 (workspace main
```

Nevertheless commands like `npm config ls -l` and `npm doctor` indicate, even this npm call still uses configuration and caches outside the ${DEVON_IDE_HOME}. I personally do not mind that - but to minimize this potential outside influence, the ng setup should be more like

```
${DEVON_IDE_HOME}/software/node/npm install --prefix ${DEVON_IDE_HOME}/software/node --global @angular/cli@latest
```

I also have removed the `--unsafe` - but just because I can't locate in the current npm documentation, what that shall control exactly.

**Comment by hohwille** [comment]
duplicate of #168

**Comment by markusschuh** [comment]
I'd say this isn't a duplicate.

The bug here is:
npm installed at DEVON_IDE_HOME\software\node\npm is not used before another `setup` run

**Comment by hohwille** [comment]
I still do not get it. Can someone explain how to reproduce the bug you are talking about?
I still assume you tested the latest beta. However, the fix is not yet released...

**Comment by hohwille** [comment]
beta12 is out:
https://repo1.maven.org/maven2/com/devonfw/tools/ide/devon-ide-scripts/3.0.0-beta12/

Can you confirm if the error still occurs with beta12?
Otherwise I will close this issue as duplicate of #168

**Comment by hohwille** [comment]
Fixed with PR #194
